sdl_android/src/main/java/com/smartdevicelink/api/screen/SoftButtonManager.java

```diff
@@ -0,0 +1,549 @@
+                }
+            });
+        }
+
+
+        // Upload other state images
+        if (otherStatesToBeUploaded.size() > 0) {
+            Log.v(TAG, "Uploading soft button other state artworks");
+            fileManager.uploadArtworks(otherStatesToBeUploaded, new MultipleFileCompletionListener() {
+                @Override
+                public void onComplete(Map<String, String> errors) {
+                    if (errors != null && errors.size() > 0) {
+                        Log.e(TAG, "Error uploading soft button artworks");
+                    }
+                    Log.d(TAG, "Soft button other state artworks uploaded");
+                    // In case our soft button states have changed in the meantime
+                    update(cachedListener);
+                }
+            });
+        }
+
+        // This is necessary because there may be no images needed to be uploaded
+        update(cachedListener);
+    }
+
+    /**
+     * Update the SoftButtonManger by sending a new Show RPC to reflect the changes
+     * @param listener a CompletionListener
+     */
+    protected void update(CompletionListener listener) {
+        cachedListener = listener;
+
+        if (batchUpdates) {
+            return;
+        }
+
+        // Don't send if we're in HMI NONE
+        if (currentHMILevel == null || currentHMILevel == HMILevel.HMI_NONE) {
+            return;
+        }
+
+        Log.v(TAG, "Updating soft buttons");
+
+        cachedListener = null;
+
+
+        // Check if we have update already in progress
+        if (inProgressShowRPC != null) {
+            Log.d(TAG, "In progress update exists, queueing update");
+            // If we already have a pending update, we're going to tell the old listener that it was superseded by a new update and then return
+            if (queuedUpdateListener != null) {
+                Log.d(TAG, "Queued update already exists, superseding previous queued update");
+                queuedUpdateListener.onComplete(false);
+                queuedUpdateListener = null;
+            }
+
+            // Note: the queued update will be started after the in-progress one finishes
+            if (listener != null) {
+                queuedUpdateListener = listener;
+            }
+            hasQueuedUpdate = true;
+            return;
+        }
+
+
+        // Send Show RPC with soft buttons representing the current state for the soft button objects
+        inProgressListener = listener;
+        inProgressShowRPC = new Show();
+        inProgressShowRPC.setMainField1(getCurrentMainField1());
+        if (softButtonObjects == null) {
+            Log.d(TAG, "Soft button objects are null, sending an empty array");
+            inProgressShowRPC.setSoftButtons(new ArrayList<SoftButton>());
+        } else if ((currentStateHasImages() && !allCurrentStateImagesAreUploaded()) && (softButtonCapabilities == null || !softButtonCapabilities.getImageSupported())) {
```

Comment:
Shouldn't this if be an OR instead of an AND?
`else if ((currentStateHasImages() && !allCurrentStateImagesAreUploaded()) || (softButtonCapabilities == null || !softButtonCapabilities.getImageSupported())) { `

Reply:
Yeah, that is right. Fixed that and also created a new private method to check if softButtonImages are supported